sdk/testing/stepwise/stepwise_test.go

```diff
@@ -0,0 +1,285 @@
+// if the VAULT_ACC environment variable is not set. This test is seperate from
+// the table tests due to the unsetting/re-setting of the environment variable,
+// which is assumed/needed for all other tests.
+func TestStepwise_Run_SkipIfNotAcc(t *testing.T) {
+	if err := os.Setenv(TestEnvVar, ""); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	defer os.Setenv(TestEnvVar, "1")
+	skipCase := Case{
+		Environment: new(mockEnvironment),
+		Steps:       []Step{Step{}},
+	}
+
+	expected := mockT{
+		SkipCalled: true,
+	}
+
+	testT := new(mockT)
+	Run(testT, skipCase)
+
+	if testT.SkipCalled != expected.SkipCalled {
+		t.Fatalf("expected SkipCalled (%t), got (%t)", expected.SkipCalled, testT.SkipCalled)
+	}
+}
+
+func TestStepwise_Run_Basic(t *testing.T) {
+	testRuns := map[string]testRun{
+		"basic": {
+			steps: []Step{
+				stepFunc("keys", ListOperation, false),
+			},
+			environment: new(mockEnvironment),
+		},
+		"error": {
+			expectedTestT: &mockT{
+				ErrorCalled: true,
+			},
+			steps: []Step{
+				stepFunc("keys", ListOperation, false),
+				stepFunc("keys", ReadOperation, true),
+			},
+			environment: new(mockEnvironment),
+		},
+		"nil-env": {
+			expectedTestT: &mockT{
+				FatalCalled: true,
+			},
+			steps: []Step{
+				stepFunc("keys", ListOperation, false),
+			},
+		},
+		"skipTeardown": {
+			steps: []Step{
+				stepFunc("keys", ListOperation, false),
+				stepFunc("keys", ReadOperation, false),
+			},
+			skipTeardown: true,
+			environment:  new(mockEnvironment),
+		},
+	}
+
+	for name, tr := range testRuns {
+		t.Run(name, func(t *testing.T) {
+			testT := new(mockT)
+			expectedT := tr.expectedTestT
+			if expectedT == nil {
+				expectedT = new(mockT)
+			}
+			testCase := Case{
+				Steps:        tr.steps,
+				SkipTeardown: tr.skipTeardown,
+			}
+
+			if tr.environment != nil {
+				testCase.Environment = tr.environment
+			}
+
+			Run(testT, testCase)
+
+			if tr.environment == nil && !testT.FatalCalled {
+				t.Fatal("expected FatalCalled with nil environment, but wasn't")
+			}
+
+			if tr.environment != nil {
+				if tr.skipTeardown && tr.environment.teardownCalled {
+					t.Fatal("SkipTeardown is true, but Teardown was called")
+				}
+				if !tr.skipTeardown && !tr.environment.teardownCalled {
+					t.Fatal("SkipTeardown is false, but Teardown was not called")
+				}
+			}
+
+			if expectedT.ErrorCalled != testT.ErrorCalled {
+				t.Fatalf("expected ErrorCalled (%t), got (%t)", expectedT.ErrorCalled, testT.ErrorCalled)
+			}
+		})
+	}
+}
+
+func TestStepwise_makeRequest(t *testing.T) {
+	me := new(mockEnvironment)
+	me.Setup()
+	testT := new(mockT)
+
+	type testRequest struct {
+		Operation         Operation
+		Path              string
+		ExpectedRequestID string
+		ExpectErr         bool
+	}
+	testRequests := map[string]testRequest{
+		"list": {
+			Operation:         ListOperation,
+			Path:              "keys",
+			ExpectedRequestID: "list-request",
+		},
+		"read": {
```

Comment:
Looks like `mockEnvironment` has a revoke endpoint but we're not testing it here. Might be also good to test create and update operations, as well as auth'ed/unauth'ed requests.

Reply:
Unless you object I'd like to defer those additions until after the merge, and the first driver is added which will bring some plugin tests that use those endpoints as well.

Reply:
Since `mockEnvironment` is not dependent on external plugins or drivers, having it here now can ensure that `makeRequest` will work properly when plugins tests do end up calling it. I'd lean towards adding those cases for completeness.

Reply:
I've updated the tests to do basic request type counting (revokes, reads, et. al.) in 8d5e1e65f , please take a look when you can!

Reply:
Looks great, thanks!